src/main/java/com/google/cloud/datafusion/plugin/successfactors/source/metadata/SuccessFactorsEntityProvider.java

```diff
@@ -0,0 +1,198 @@
+    this.edmMetadata = edmMetadata;
+  }
+
+  /**
+   * Find and return the EdmEntitySet instance from the metadata.
+   *
+   * @param entityName SuccessFactors entity name
+   * @return instance of EdmEntitySet class
+   * @throws EdmException Expected exception
+   */
+  @Nullable
+  public EdmEntitySet getEntitySet(String entityName) throws EdmException {
+    if (SuccessFactorsUtil.isNotNullOrEmpty(entityName)) {
+      for (EdmEntitySet edmEntitySet : edmMetadata.getEntitySets()) {
+        if (edmEntitySet.getName().equals(entityName)) {
+          return edmEntitySet;
+        }
+      }
+    }
+
+    String debugMsg = String.format("Entity name: '%s', could not find EntitySet for the given entity name. " +
+      "Root cause: null / empty or invalid entity name is provided.", entityName);
+    LOGGER.debug(debugMsg);
+
+    return null;
+  }
+
+  @Nullable
+  public EdmEntityType getEntityType(String entityName) throws EdmException {
+    EdmEntitySet entitySet = getEntitySet(entityName);
+    if (entitySet != null) {
+      return entitySet.getEntityType();
+    }
+
+    String debugMsg = String.format("Entity name: '%s', could not find EntityType for the given entity name. " +
+      "Root cause: null / empty or invalid entity name is provided.", entityName);
+    LOGGER.debug(debugMsg);
+
+    return null;
+  }
+
+  /**
+   * Get list of all the default property name associated with the give 'entityName'
+   *
+   * @param entityType service entity type
+   * @return list of default property name
+   * @throws EdmException Expected exception
+   */
+  @Nullable
+  public List<String> getEntityPropertyList(EdmEntityType entityType) throws EdmException {
+    if (entityType != null) {
+      return entityType.getPropertyNames();
+    }
+    return null;
+  }
+
+  /**
+   * Returns the list of default entity from the metadata.
+   *
+   * @return list of default entity from the metadata
+   * @throws EdmException Expected exception
+   */
+  public List<EdmEntitySet> getDefaultEntitySet() throws EdmException {
+    return edmMetadata.getDefaultEntityContainer().getEntitySets();
+  }
+
+  /**
+   * Find and return the last navigation property from the given navigation path.
+   *
+   * @param entityName service entity name
+   * @param navPath    can have navigation path or navigation property name
+   * @return returns the relevant value
+   * @throws EdmException Expected exception
+   */
+  @Nullable
+  public EdmNavigationPropertyImplProv getNavigationProperty(String entityName, String navPath)
+    throws EdmException {
+
+    if (SuccessFactorsUtil.isNotNullOrEmpty(entityName) && SuccessFactorsUtil.isNotNullOrEmpty(navPath)) {
+      EdmEntitySet entitySet = getEntitySet(entityName);
+      if (entitySet != null) {
+        EdmEntityType entityType = entitySet.getEntityType();
+        EdmNavigationPropertyImplProv association = null;
+
+        String[] navNames = navPath.split("/");
```

Comment:
If the navigation path does not have `/` it will lead to an NPE in the next line.
Are we checking this somewhere upstream?